### DISCUSSION TITLE
명령어 파이프라이닝과 해저드

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#259] [Computer Architecture]
.

**Comment by 3Juhwan** [comment]
## 명령어 파이프라인
명령어 파이프라인은 여러 명령어를 동시에 처리하기 위해 각 명령어를 여러 단계로 분할하고, 각 단계를 다른 명령어와 겹쳐서 실행하는 방법이다. 전통적인 CPU는 한 번에 하나의 명령어를 처리하는 반면, 파이프라인 기법을 사용하면 여러 명령어를 각기 다른 단계에서 동시에 처리할 수 있다. 

하나의 명령어는 여러 단계에 걸쳐 실행된다. 각 단계는 비슷한 시간 간격으로 나눌 수 있다. 주로 다음 4단계를 거치며 CPU에 따라 다르다. 
```text
명령어 인출 -> 명령어 해석 -> 명령어 실행 -> 결과 저장
```

<div align="center">
  <img src="https://github.com/user-attachments/assets/d6025c6a-40d7-45f2-ba2b-c87c6b320aca" alt="image" width="300">
</div>
</br>


CPU는 **각 단계를 동시에 실행할 수 있다**. 명령어 A를 인출하는 동시에 명령어 B를 해석하고, 명령어 C를 실행할 수 있다. 이처럼 명령어를 겹쳐서 실행하면 순차 실행보다 효율적이다. 

## 파이프라인 위험(해저드)

CPU의 명령어 파이프라인에서 성능 저하를 일으키는 문제를 말한다. 파이프라인 해저드는 파이프라인이 최적의 성능으로 실행되지 못하게 만드는 상황을 일으키며, 주로 명령어 간의 종속성이나 자원 충돌에서 발생한다.

**1. 데이터 위험**

데이터 의존성에 의해 발생한다. 아래 상황에서 명령어 B는 명령어 A의 연산 결과인 R1에 의존하고 있다. 명령어 A의 결과 저장 단계가 완료되어야 명령어 B는 R1을 사용할 수 있다. 
```text
명령어 A: R1 <- R2 + R3
명령어 B: R4 <- R1 + R5
```

**2. 제어 위험**

분기(branch)나 조건문 등으로 인해 다음에 실행할 명령어가 무엇인지 결정되지 않아 파이프라인이 멈추게 되는 상황이다. 분기 예측 기법이 사용되기도 하지만, 예측에 실패하면 계산된 파이프라인은 모두 버려야 해서 성능 저하가 발생한다. 

**3. 구조적 위험**

서로 다른 명령어가 동시에 ALU, 레지스터와 같은 CPU 자원을 사용하려 할 때 발생한다. 한 명령어가 자원을 사용하면 다른 명령어는 멈추게 되어 파이프라인은 동작을 멈추고 성능 저하가 발생한다.

**Comment by le2sky** [comment]
명령어 파이프라이닝(instruction pipelining)은 명령어를 실행하는데 사용되는 하드웨어를 여러 개의 독립적인 단계(stage)로 분할하고, 각 하드웨어를 동시에 서로 다른 명령어들을 처리하도록 함으로써 CPU의 성능을 높여주는 기술을 말한다. 이때, 명령어 파이프라인의 분할 단계가 많아질수록 처리 속도가 높아진다.

**2-단계 명령어 파이프라인**

<div align = "center">
    <img height = 300 width = 500 src ="https://github.com/user-attachments/assets/95f8c20b-2186-4506-be7b-b3aaedeebb7b" />
</div>

- 각 사이클(인출, 실행)의 동작을 처리하는 하드웨어를 독립적인 모듈로 구성할 수 있다면, 각 모듈이 서로 다른 명령어를 동시에 처리할 수 있다. 이 경우, 인출 단계(fetch stage)와 실행 단계(execute stage)라는 두 개의 파이프라인 단계들로 분리하여 구성한다.
- 두 파이프라인 단계들에 하나의 클록 신호를 동시에 인가한다면 해당 단계들의 동작 시간을 일치시킬 수 있다.
- 각 단계는 서로 다른 명령어에 대하여 각자의 동작을 수행할 수 있게 된다.
- 위 그림처럼 다음에 실행될 명령어를 미리 인출하는 것을 **명령어 선인출(instruction prefetch)** 혹은 **인출 중복(fetch overlap)** 이라고 한다.
- 하지만,  명령어의 인출과 실행에 같은 길이의 시간이 소요되는 경우에만 속도 향상을 얻을 수 있다. 
    - 실제 명령어 처리 과정을 보면 일반적으로 실행 단계에서 소요되는 시간이 인출 단계보다 길다.
    - 따라서 인출 단계는 하나의 명령어를 인출한 다음에 즉시 다음 명령어 인출을 시작하지 못하고, 실행 단계의 동작이 끝날 때까지 기다려야 한다.
    - 2-단계 파이프라인을 이용하면 속도 향상을 얻을 수 있지만, 위와 같은 이유때문에 실제로 2배만큼 빨라지지는 못한다.

**4-단계 명령어 파이프라인**

파이프라인 단계들의 처리 시간이 동일하지 않음으로 인해 발생하는 효율저하를 방지하기 위해서는 단계를 더 많이 분할하여 단계 별 처리 시간을 거의 같아지도록 하는 것이다. 2-단계 파이프라인의 실행 단계를 세부적 동작에 따라 분할하여 다음과 같이 구성하는 경우를 생각해보자. 아래와 같이 4-단계 명령어 파이프라인으로 구성하면 각 단계에서 걸리는 시간들이 거의 같아질 수 있다.

- 명령어 인출(IF) : 기억장치로부터 명령어 인출
- 명령어 해독(ID) : 명령어 해독
- 오퍼랜드 인출(OF) : 기억장치로부터 오퍼랜드 인출
- 실행(EX) : 지정된 연산을 수행하고, 결과를 저장

<div align = "center">
    <img height = 300 width = 500 src ="https://github.com/user-attachments/assets/f07bed06-9e2f-4c58-a10e-62c2fc660d63" />
</div>

**파이프라인 위험(pipeline hazard)**
파이프라이닝은 높은 성능을 가져오기는 하지만, 때로는 성능 향상에 실패하기도 한다. 이를 파이프라인 위험(pipeline hazard)이라고 한다.

- 데이터 위험(data hazard) : 
    - 명령어 간 데이터 의존성에 의해 발생한다. 모든 명령어를 동시에 처리할 수 없는데, 어떤 명령어는 이전 명령어를 끝까지 실행해야만 비로소 실행 가능하다.
    - 이처럼 데이터 의존적인 두 명령어를 동시에 실행하려고 하면 파이프라인이 제대로 작동하지 않는 것을 데이터 위험이라고 한다.
- 제어 위험(control hazard) :
    - 주로 분기 등으로 인한 프로그램 카운터의 갑작스러운 변화에 의해 발생한다.
    - 선인출한 명령어가 프로그램 카운터의 갑작스러운 변화에 의해 필요없어진다. 이를 제어 위험이라고 한다.
    - 이를 해결하기 위해 사용하는 것을 분기 예측(branch prediction)이라고 한다.
    - 이는 프로그램이 어디로 분기할지 미리 예측하여 주소를 인출하는 기술이다.
- 구조적 위험(structural hazard)
    - 명령어들을 겹쳐 실행하는 과정에서 서로 다른 명령어가 동시에 ALU, 레지스터와 같은 CPU 부품을 사용하려고 할 때 발생한다.
    - 자원 위험(resource hazard)이라고도 한다.

각 위험의 해결 방안은 [이곳](https://80000coding.oopy.io/e20ff4ae-ee5a-40dc-9c23-13a369e296d6)에서 더욱 자세하게 볼 것
